### PR TITLE
`CustomRecurrentLayer` arbitrary depth layer stacks

### PR DESCRIPTION
**Pull request by Britefury**
This PR addresses issue #777, allowing the input-to-hidden and hidden-to-hidden layer stacks to have arbitrary depth (with unit tests).

### CODE REVIEW

**Comment by f0k** [comment]
Good spot! But if I see correctly, the actual problem seems to be this line only: https://github.com/Lasagne/Lasagne/blob/0abf57f/lasagne/layers/recurrent.py#L299
Can't we fix this more easily by just computing the input shape of `hidden_to_hidden` in a correct way, without adding more attributes and changing `get_output_for`?

**Comment by Britefury** [comment]
Good call. That additional code was in there because I also attempted to allow the input-to-hidden and hidden-to-hidden graphs to have arbitrary structure, including branching back to the input graph. This turned out not to work as it was a bad idea.

There is still a little additional sanity checking code at the beginning of the constructor that checks that the input-to-hidden and hidden-to-hidden graphs contain a unique `InputLayer`, just in case anyone tries anything unwise. It can be a bugger to find the root cause if you do...

File: lasagne/layers/recurrent.py [hunk 1]
```diff
@@ -222,6 +222,23 @@ def __init__(self, incoming, input_to_hidden, hidden_to_hidden,
 
         super(CustomRecurrentLayer, self).__init__(incomings, **kwargs)
 
+        input_to_hidden_layers = helper.get_all_layers(input_to_hidden)
+        input_to_hidden_in_lyrs = [lyr for lyr in input_to_hidden_layers
+                                   if isinstance(lyr, InputLayer)]
```
Comment (f0k) [inline lasagne/layers/recurrent.py]: Can you combine this into one call? The fewer local variables are initialized, the less I need to worry about when reading the code. I'd also prefer not to use abbreviations:
```python
input_to_hidden_input_layers = [layer for layer
                                in helper.get_all_layers(input_to_hidden)
                                if isinstance(layer, InputLayer)]
```

File: lasagne/layers/recurrent.py [hunk 2]
```diff
@@ -222,6 +222,23 @@ def __init__(self, incoming, input_to_hidden, hidden_to_hidden,
...
+        if len(input_to_hidden_in_lyrs) != 1:
+            raise ValueError('input_to_hidden_in_layer '
+                             'not provided and input_to_hidden does not '
+                             'have a unique InputLayer')
```
Comment (f0k) [inline lasagne/layers/recurrent.py]: This indicates you could provide the input layer in some way, but you can't. I'd say:
```
`input_to_hidden` must have exactly one InputLayer, but it has %d
```

File: lasagne/layers/recurrent.py [hunk 3]
```diff
@@ -294,16 +311,16 @@ def __init__(self, incoming, input_to_hidden, hidden_to_hidden,
...
-                             "input_to_hidden.output_shape={} and "
-                             "hidden_to_hidden.input_shape={}".format(
-                                 input_to_hidden.output_shape,
-                                 hidden_to_hidden.input_shape))
+                                     h_to_h_input_shape[1:])):
+            raise ValueError(
+                "The output shape for input_to_hidden must be equal to the "
+                "input shape of hidden_to_hidden after the first dimension, "
+                "but input_to_hidden.output_shape={} and "
+                "hidden_to_hidden:input_layer.output_shape={}".format(
```
Comment (f0k) [inline lasagne/layers/recurrent.py]: Maybe `hidden_to_hidden:input_layer.shape` is easier to grasp? I don't know. `InputLayer` has a `shape` property that may feel more natural than `output_shape`.

File: lasagne/tests/layers/test_recurrent.py [hunk 4]
```diff
@@ -189,7 +189,129 @@ def test_custom_recurrent_arbitrary_shape():
...
+
+    out = theano.function([l_in.input_var], lasagne.layers.get_output(l_rec))
+    out_shape = out(np.zeros((n_batch, n_steps, n_channels, width, height),
+                             dtype=theano.config.floatX)).shape
+    assert out_shape == (n_batch, n_steps, n_out_filters, width, height)
+
+
+def test_custom_recurrent_non_unique_inputs():
+    # Check that the custom recurrent layer can handle a hidden-to-hidden
+    # network with an arbitrary depth
```
Comment (f0k) [inline lasagne/tests/layers/test_recurrent.py]: This doesn't seem like the correct comment for this test.

**Comment by Britefury** [comment]
@f0k I have addressed your comments and change requests; I hope its in a good enough state now.

**Comment by f0k** [comment]
Looks good, thank you! Can you squash into a single commit again? We're trying to have a clean history. Cheers!

**Comment by Britefury** [comment]
Done.

**Comment by f0k** [comment]
Merged, thanks again!

**Pull request merged**